### ISSUE TITLE
multiple calls to subst with the same inputs returns different outcomes

### ISSUE THREAD

**Issue by demanuel** [rakudo/rakudo#5154]
I was trying to get a regexp that would remove quotes from the begining/end of a string.

![imagem](https://user-images.githubusercontent.com/1687786/212176031-b4721882-b1d9-4e45-b4a6-15f84ea1d072.png)

I have no clue what's going on here, but same code returns different outcomes.

```
rakudo --version
Welcome to Rakudo™ v2022.12-944-g1b5b57514.
Implementing the Raku® Programming Language v6.d.
Built on MoarVM version 2022.12-14-gebefe2618.

```

Code:
```
my $sentence = '"tag que começa com aspas"';
say "0 "~$sentence.subst(/^ '"' (.*?) '"' $/, $0);

say "1 " ~ $sentence;

say "2 " ~ $sentence.subst(/^ '"' (.*?) '"' $/, $0, :g);

say "3 " ~ $sentence;

say "4 " ~ $sentence.subst(/^ '"' (.*?) '"' $/, $0, :g);

say "5 " ~ $sentence;

say "6 " ~ $sentence.subst(/^ '"' (.*?) '"' $/, $0, :g);

say "7 " ~ $sentence.subst(/^ '"' (.*?) '"' $/, ~$0, :g);

say "8 " ~ $sentence.subst(/^ '"' (.*?) '"' $/, ~$0, :g);

say "9 " ~ $sentence.subst(/^ '"' (.*?) '"' $/, ~$0);

say "10 " ~ $sentence.subst(/^ '"' (.*?) '"' $/, $0);

say "11 " ~ $sentence.subst(/^ '"' (.*?) '"' $/, $0, :g);

say "12 " ~ $sentence;
```

Output:
```
> raku bug.raku
Use of Nil in string context
  in block <unit> at bug.raku line 2
0 
1 "tag que começa com aspas"
2 tag que começa com aspas
3 "tag que começa com aspas"
4 "tag que começa com aspas"
5 "tag que começa com aspas"
6 "tag que começa com aspas"
7 "tag que começa com aspas"
8 "tag que começa com aspas"
9 "tag que começa com aspas"
10 tag que começa com aspas
11 tag que começa com aspas
12 "tag que começa com aspas"
```

**Comment by pmichaud** [comment]
I suspect the problem is misunderstanding the meaning of ```$0``` in these statements.   I think that passing $0 as an argument means it's the result of the first capture of the previous match (or whatever the :g equivalent is).  

Thus the first time .subst is invoked (first statement[1]), $0 is not set and you get the "Use of Nil in string context" error.  

The second time it's invoked ([2]), $0 is referring to whatever was matched in the previous .subst statement, which is the text between the two double-quotes.  So, the string with double-quotes gets replaced by $0 from the .subst in [1] which doesn't have the double-quotes, and that's the result.  As a side effect, the :g flag in [2] means that $0 for subsequent statements is a match object containing a subcapture of matches from the .subst -- this is different from what happened in [1] when $0 was left as a singular Match object.

So, the third time .subst is invoked ([4]), $0 is a match with a subcapture which ends up stringifying to ```"tag que começa com aspas"``` (including the double-quotes), and so the full string gets replaced with a version that has double-quotes (i.e. it doesn't appear to change).  This happens for many of the following statements.

In short, the confusion comes down to two items:
 - $0 in the call to .subst is referring to the results of a _previous match_ , not the current match being performed
 - The presence or absence of ```:g``` in the .subst calls changes what $0 contains heading into the next call to .subst .
 
 That's my guess, although I'm a bit rusty on Raku stuff these days.

Pm

**Comment by demanuel** [comment]
That makes sense. I think you are right and if so, IMHO, the behaviour of $0 is not consistent:

```
$ raku
Welcome to Rakudo™ v2022.12.
Implementing the Raku® Programming Language v6.d.
Built on MoarVM version 2022.12.

To exit type 'exit' or '^D'
[0] > my $sentence = '"tag que começa com aspas"'
"tag que começa com aspas"
[1] > $sentence ~~ s/^ '"' (.*?) '"'$/$0/;
｢"tag que começa com aspas"｣
 0 => ｢tag que começa com aspas｣
[2] > $sentence
tag que começa com aspas
[3] > 

```

**Comment by vrurg** [comment]
Here you misunderstand the order of things. `s///` is a single clause which both sets and uses `$0`. When you invoke `.subst` method you pass two arguments: a regex and a variable `$0`. Since regex evaluation happens later, inside the method, it is using, roughly, method's own lexical `$0`.

Just keep in mind that `$0` is not a global, its a lexical.

**Comment by lizmat** [comment]
You **CAN** refer to captures in the replacement part, you just have to make sure that they have their own closure:
````
say "foobar".subst(/ (.) $0 /, { $0.Str.succ });  # fpbar
```
Aka, look at 2 consecutive identitcal characters, and then replace that by the next alphabetical character.  In this case:  `oo` -> `p`.

You need the `{  }` to tell `.subst` basically to not use the `$0` from the scope of where the `.subst` is called.

Hope this makes sense  :-)

**Comment by demanuel** [comment]
@lizmat Super! I didn't know about that. Thank you very much!
